pkg/reconciler/workload/placement/placement_reconcile_scheduling.go

```diff
@@ -78,19 +114,17 @@ func (r *placementSchedulingReconciler) reconcile(ctx context.Context, placement
 	// TODO(qiujian16): we currently schedule each in each location independently. It cannot guarantee 1 cluster is scheduled per location
 	// when the same synctargets are in multiple locations, we need to rethink whether we need a better algorithm or we need location
 	// to be exclusive.
-	if len(syncTargets) > 0 {
-		scheduledSyncTarget := syncTargets[rand.Intn(len(syncTargets))]
-		expectedAnnotations[workloadv1alpha1.InternalSyncTargetPlacementAnnotationKey] = workloadv1alpha1.ToSyncTargetKey(syncTargetClusterName, scheduledSyncTarget.Name)
-		updated, err := r.patchPlacementAnnotation(ctx, clusterName, placement, expectedAnnotations)
-		return reconcileStatusContinue, updated, err
-	}
-
-	return reconcileStatusContinue, placement, nil
+	scheduledSyncTarget := result.validSyncTargets[rand.Intn(len(result.validSyncTargets))]
+	expectedAnnotations[workloadv1alpha1.InternalSyncTargetPlacementAnnotationKey] = workloadv1alpha1.ToSyncTargetKey(result.locationWorkspace, scheduledSyncTarget.Name)
+	updated, err := r.patchPlacementAnnotation(ctx, clusterName, placement, expectedAnnotations)
+	return reconcileStatusContinue, updated, err
 }
 
-func (r *placementSchedulingReconciler) getAllValidSyncTargetsForPlacement(clusterName logicalcluster.Name, placement *schedulingv1alpha1.Placement) (logicalcluster.Name, []*workloadv1alpha1.SyncTarget, error) {
+func (r *placementSchedulingReconciler) getAllValidSyncTargetsForPlacement(ctx context.Context, clusterName logicalcluster.Name, placement *schedulingv1alpha1.Placement) *schedulingResult {
```

Comment:
embedding an error is very unusual in Golang. Why is this done?

Reply:
schedulingResult is a very strange name for what it is: `getAllValidSyncTargetsForPlacement`. Either all targets, or a result of scheduling. Cannot be both.